### ISSUE TITLE
saved.shift is not a function

### ISSUE THREAD

**Issue by dsandber** [jquery/jquery#4350]
My app auto-upgraded to 3.4.0 from 3.3.1 and broke with this error:

saved.shift is not a function

I'll append the browser version in a comment shortly.

**Comment by dsandber** [comment]
user_agent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36"

trace: "TypeError: saved.shift is not a function
    at HTMLDivElement.handler (https://inbox-credit.vertical.arbcalls.com/js/vue/app.aaa76e4c45a8eb6f83bf.js:10492:27)
    at HTMLDivElement.dispatch (https://inbox-credit.vertical.arbcalls.com/js/vue/app.aaa76e4c45a8eb6f83bf.js:10244:27)
    at HTMLDivElement.elemData.handle (https://inbox-credit.vertical.arbcalls.com/js/vue/app.aaa76e4c45a8eb6f83bf.js:10051:28)

**Comment by timmywil** [comment]
Sorry, there is not enough information to go on here. You can find help on https://stackoverflow.com. If after sorting out the issue you find there is a change needed in jQuery core, please open a new issue and follow the instructions in the issue template, including providing a test case from https://jsbin.com.

**Comment by dsandber** [comment]
Sorry, because it wasn't triggered directly by my code, and because it doesn't seem to happen consistently, I don't know how to create a test case.

**Comment by timmywil** [comment]
Fair enough, but the issue needs to be fleshed out more regardless.

**Comment by dsandber** [comment]
Luckily "saved.shift" only occurs once in the source-code.  Strange that the comment says IE <= 9 - 11+ but that it is being run on Chrome.


```
dataPriv.set( this, type, jQuery.event.trigger(
  // Support: IE <=9 - 11+
  // Extend with the prototype to reset the above stopImmediatePropagation()
  jQuery.extend( saved.shift(), jQuery.Event.prototype ),
  saved,
  this
) );
```

**Comment by mgol** [comment]
@dsandber Unfortunately, we cannot do much without a test case we can run, the error itself doesn't say much. We usually require a small test case without external dependencies as most errors user have is actually a bug on their app part but since this code is new in 3.4.0 & it's one of the most trickiest new part, it's possible we broke something so I'd be open to looking at a slightly bigger test case. But I still need at least a website with this bug, with an easy way to reproduce the bug. Ideally I'd have both one with jQuery 3.3.1 & 3.4.0 to compare but I could try with just 3.4.0 as well if the older one is not available.

**Comment by dsandber** [comment]
Yeah, I understand -- probably best to wait until more people report this error and hopefully someone can develop a simple test case.  I literally wouldn't know where to start since the stack trace doesn't include my code, the error doesn't happen consistently, and I don't understand what the code I quoted does.

**Comment by fancyapps** [comment]
I noticed this error message after upgrading to 3.4.0., too. I could not figure out when/why, I`ll try to create a minimal example.

**Comment by XhmikosR** [comment]
@mgol we hit this in Bootstrap 4.x https://travis-ci.org/twbs/bootstrap/jobs/519620443

**Comment by dsandber** [comment]
Given three independent reports, perhaps the bug should be re-opened to make it easier for others to find? (since closed bugs aren't searched by default)

**Comment by dmethvin** [comment]
Reopening it is fine, but we still don't have a test case. Can we at least get a stack trace?

**Comment by XhmikosR** [comment]
The stacktrace is shown in the Travis build, but it's using the minified build. AFAICT it only affects IE 10, at least in our test suite.

**Comment by dsandber** [comment]
> Can we at least get a stack trace?

My second comment includes a stack trace.  

> AFAICT it only affects IE 10, 

Chrome too.

**Comment by XhmikosR** [comment]
I also pushed a patch with the non-minified build here: https://travis-ci.org/twbs/bootstrap/jobs/519635723

**Comment by dmethvin** [comment]
I was thinking those stack traces were truncated, but they're complete (from the point of event invocation) but just not informative. *Some* code is triggering an event of *some* type and it causes an error. If we could fill in the two "some" with specifics we could diagnose the issue.

**Comment by XhmikosR** [comment]
One test that fails is doing the following:

```js
        // Key escape
        $input.trigger('focus').trigger($.Event('keydown', {
          which: 27
        }))
        assert.ok(!$dropdown.parent('.dropdown').hasClass('show'), 'dropdown menu is not shown')
        done()
```

BTW I'm also getting a Symbol is not a function error on IE 10 with 3.4.0.

**Comment by gibson042** [comment]
As @dsandber observed, the source text occurs [only once](https://github.com/jquery/jquery/blob/fe5f04de8fde9c69ed48283b99280aa6df3795c7/src/event.js#L601) in jQuery, in a context like this:
```js
	𝘀𝗮𝘃𝗲𝗱 = 𝗱𝗮𝘁𝗮𝗣𝗿𝗶𝘃.𝗴𝗲𝘁( 𝘁𝗵𝗶𝘀, 𝘁𝘆𝗽𝗲 );

if ( ( event.isTrigger & 1 ) && this[ type ] ) {

	// Interrupt processing of the outer synthetic .trigger()ed event
	if ( !saved ) {

		// Store arguments for use when handling the inner native event
		saved = slice.call( arguments );
		𝗱𝗮𝘁𝗮𝗣𝗿𝗶𝘃.𝘀𝗲𝘁( 𝘁𝗵𝗶𝘀, 𝘁𝘆𝗽𝗲, 𝘀𝗮𝘃𝗲𝗱 );
		…
	}
	…

// If this is a native event triggered above, everything is now in order
// Fire an inner synthetic event with the original arguments
} else if ( saved ) {

	// ...and capture the result
	dataPriv.set( this, type, jQuery.event.trigger(

		// Support: IE <=9 - 11+
		// Extend with the prototype to reset the above stopImmediatePropagation()
		jQuery.extend( 𝘀𝗮𝘃𝗲𝗱.𝘀𝗵𝗶𝗳𝘁(), jQuery.Event.prototype ),
		saved,
		this
	) );
```
`type` comes from [leverageNative](https://github.com/jquery/jquery/blob/fe5f04de8fde9c69ed48283b99280aa6df3795c7/src/event.js#L538) as either "click", "focus", or "blur", and `dataPriv` is a [Data](https://github.com/jquery/jquery/blob/fe5f04de8fde9c69ed48283b99280aa6df3795c7/src/data/Data.js) map that _should_ be library-internal but is actually exposed to consumer code via [`jQuery._data`](https://github.com/jquery/jquery/blob/fe5f04de8fde9c69ed48283b99280aa6df3795c7/src/data.js#L85-L89).

We expect `dataPriv.get( this, type )` to be fully under the control of leverageNative, and thus either `false` or an array of trigger data set in outermost event handling. But if some _other_ code is using `jQuery._data` to set e.g. "focus" data to a non-`false` non-array value, then we would see precisely this kind of error—and I strongly suspect that's what's going on. We [specifically addressed this point](https://github.com/jquery/jquery/pull/4279#discussion_r250234565) before landing the new code, but obviously can only protect against _internal_ collisions rather than external.

We need to decide whether to rename the private data key or leave this as a purely consumer issue.
Can everyone experiencing problems please find the use of `jQuery._data` that messes with "focus" and/or "blur" and/or "click"?

**Comment by fancyapps** [comment]
Hi! This simple test case fails using IE11 after clicking link the second time:

```HTML
<p>
  <a href="#" id="test-ie11">Click me</a>
</p>

<script>
  $('#test-ie11').click(function(e) {
    $(this).trigger("blur")
    return false;
  });
</script>
</script>
```

Tracing does not give much info (same as in previous comments), but `saved` variable is Boolean (true) instead of object.

**Comment by gibson042** [comment]
Thanks so much @fancyapps, this is real and I feel pretty dumb right now. The issue is that in environments where native focus/blur are async, the leverageNative native event handling still sets private data, which results in an error on the _next_ triggered invocation: https://jsbin.com/mizironoha

I think a fix needs to update detection of outer synthetic events to use not just `!saved` but also absence of sentinel data that can never be returned by the `jQuery.event.trigger` in native event handling. Before optimizing for size, I'm thinking something like https://jsbin.com/jozeromequ (in which the `// UPDATE` lines demonstrate prepending `leverageNative` itself to the saved array for this purpose).

And I also noticed in passing that the focus/blur trigger handlers call `leverageNative` every time rather than guarding it with a `dataPriv.get` call as the click trigger handler does, so we should fix that as well.

**Comment by mgol** [comment]
I added this to the newly created `3.4.1` milestone.

**Comment by ewolfman** [comment]
Hi,
When is the 3.4.1 release date?

**Comment by timmywil** [comment]
There's no set date, but it will probably be next week or the week after.

**Comment by JimmyChou0630** [comment]
hi thanks

**Comment by dmnd** [comment]
I just broke production because of this bug.

GitHub currently pushes recommendations to all repos using jquery <3.4.0 to upgrade due to CVE-2019-5428. I saw that, so I checked out the [release blog post](https://blog.jquery.com/2019/04/10/jquery-3-4-0-released/) which currently says:

> There should be no compatibility issues if upgrading from jQuery 3.0+.

So after some brief sanity testing (which seemed fine) I deployed and then errors started arriving.

Given this bug, I recommend you change the wording on the blog post for 3.4.0.

**Comment by mgol** [comment]
Fixed by PR #4354 (commit 24d71ac70406f522fc1b09bf7c4025251ec3aee6)

**Comment by vincentwoo** [comment]
Any word on when a 3.4.1 release will be cut?

**Comment by jcubic** [comment]
@vincentwoo I'm also interested when it will be out, It seems that only iOS 10 test are waiting https://github.com/jquery/jquery/milestone/20

**Comment by timmywil** [comment]
jQuery 3.4.1 has been released! https://blog.jquery.com/2019/05/01/jquery-3-4-1-triggering-focus-events-in-ie-and-finding-root-elements-in-ios-10/
